proposals/nnnn-structured-concurrency.md

```diff
@@ -148,39 +148,68 @@ In some situations the priority of a task must be escalated in order to avoid a
-
-Because the main body of the function executes concurrently with its child tasks, it is possible that `makeDinner` will reach the point where it needs the value of an `async let` (say, `veggies`) before that value has been produced. To account for that, reading a variable defined by an `async let` is treated as a potential suspension point, and therefore must be marked with `await`. When the expression on right-hand side of the `=` of an `async let` can throw an error, that thrown error can be observed when reading the variable, and therefore must be marked with some form of `try`.
-The task will suspend until the child task has completed initialization of the variable (or thrown an error), and then resume.
-
-One can think of `async let` as introducing a (hidden) future, which is created at the point of declaration of the `async let` and whose value is retrieved at the `await`. In this sense, `async let` is syntactic sugar to futures.
-
-However, child tasks in the proposed structured-concurrency model are (intentionally) more restricted than general-purpose futures. Unlike in a typical futures implementation, a child task does not persist beyond the scope in which it was created. By the time the scope exits, the child task must either have completed, or it will be implicitly awaited. When the scope exits via a thrown error, the child task will be implicitly cancelled before it is awaited.
+  // Prepare some variables to receive results from our concurrent child tasks
+  var veggies: [Vegetable]?
+  var meat: Meat?
+  var oven: Oven?
 
-Bringing it back to our example, note that the `chopVegetables()` function might throw an error if, say, there is an incident with the kitchen knife. That thrown error completes the child task for chopping the vegetables. The error will then be propagated out of the `makeDinner()` function, as expected. On exiting the body of the `makeDinner()` function with this error, any child tasks that have not yet completed (marinating the meat or preheating the oven, maybe both) will be automatically cancelled.
-
-### Task groups
-
-A *task group* defines a scope in which one can create new child tasks programmatically. As with all child tasks, the child tasks within the task group scope must complete when the scope exits, and will be implicitly cancelled first if the scope exits with a thrown error. This is equivalent to the behavior of `async let` variables, but is more suitable for a dynamic set of child tasks.
+  // Create a task group to scope the lifetime of our three child tasks
+  try await Task.withGroup(resultType: Void.self) { group in
+    await group.add {
+      veggies = try await chopVegetables()
```

Comment:
Interesting. This implementation will cause an error under the design in [Preventing Data Races in the Swift Concurrency Model](https://gist.github.com/DougGregor/10db898093ce33694139d1dcd7da3397), because the closure is modifying `veggies` in a closure that appears to run concurrently with the main body... even though this isn't actually the case. The formulation of this example in the prior revision of the document used an enum result type from the tasks so that modification of `veggies`, `meat`, and `oven` happened synchronously. I lean toward wanting the formulation from the prior revision, because it is probably the safer way in general for folks to create child tasks---make them functional, and do the mutation in the main closure in response to `next()`---but I'd love your thoughts here.

I wonder if a better first example would be `chopVegetables()`, because task groups are a whole lot more natural when working with dynamic work, and the `makeDinner()` example feels a bit contorted. It's far more amenable to `async let`.

Reply:
Now that I've written it, I'm losing confidence in my first statement: the fact that the closure passed to `withGroup` is non-concurrency/non-escaping might actually prevent the error. The comment about `chopVegetables()` potentially being a better example still stands, though

Reply:
Hopefully that's sufficient. It would be unfortunate if the data race model prevented users from taking advantage of the scoped nature of task groups to be able to make ad-hoc modifications to their outer scope. Even with `async let` sugar for the common case, it seems generally useful.